### PR TITLE
Hd sheeza

### PR DESCRIPTION
**Pull request by gem-sheeza-bakshi**

### CODE REVIEW

File: src/test/java/com/qa/helpdesk/Locators/locator.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+import org.openqa.selenium.WebElement;
+
+import java.util.List;
+
+public class locator {
+
+    // Login Page
+    public static By loginSSO = By.xpath("//button[text()= \"Login with SSO\"]");
+    public static By SignInPageHelpdesk = By.xpath("//div/img[@alt=\"helpdesk\"]");
+    public static By SignInIcon = By.xpath("//div[text()='Sign in']");
```
Comment (gem-rushilmatta) [inline src/test/java/com/qa/helpdesk/Locators/locator.java:13]: Camel Casing in variable names

File: src/test/java/com/qa/helpdesk/stepDefinition/stepDefinition.java [hunk 2]
```diff
@@ -0,0 +1,468 @@
...
+            System.out.println(e);
+        }
+        String Password = login.getProperty("password");
+        System.out.println(Password);
+        DriverAction.typeText(locator.passwordInputBox,"Password","Password entered",Password);
+        waitSec(5);
+        DriverAction.click(locator.nextButton);
+        DriverAction.waitSec(3);
+    }
+
```
Comment (gem-rushilmatta) [inline src/test/java/com/qa/helpdesk/stepDefinition/stepDefinition.java:87]: Use try Catch Block in these methods , dont use throws Exception here. If the steps gets passed without throwing error Use GemTestReporter to set Status of the Step to Passed, and if any exception occurs Use GemTestReporter to send Status as failed

File: src/test/java/com/qa/helpdesk/Locators/locator.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+    public static By emailInputBox = By.xpath("//input[@type=\"email\"]");
+    public static By nextButton = By.xpath("//input[@type=\"submit\"]");
+    public static By passwordInputBox = By.xpath("//input[@name=\"passwd\"]");
+    public static By staySignedIn = By.xpath("//input[@type=\"button\"]");
+
```
Comment (gem-rushilmatta) [inline src/test/java/com/qa/helpdesk/Locators/locator.java:19]: Use common function for similar xpath

File: src/test/java/com/qa/helpdesk/Locators/locator.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    public static By filterClosed = By.xpath("//button[text()='Closed']");
+    public static By clearBtn = By.xpath("//button[text()='Clear all']");
+
+
+    //Create Ticket
+    public static By createTicketBtn = By.xpath("//button[text()='Create New Ticket']");
+    public static By ticketCreationDetails = By.xpath("//label[contains(@class,'labelText')]");
+    public static By previewBtn = By.xpath("//button[text()='Preview']");
+    public static By submitBtn = By.xpath("//button[text()='Submit']");
+    public static String confirmButtons = "//button[text()='replace']";
```
Comment (gem-rushilmatta) [inline src/test/java/com/qa/helpdesk/Locators/locator.java:48]: Use common function for similar xpath

File: src/test/java/com/qa/helpdesk/stepDefinition/stepDefinition.java [hunk 5]
```diff
@@ -0,0 +1,468 @@
...
+
+    @And("Click on No button")
+    public void clickOnNoButton() {
+        DriverAction.click(locator.staySignedIn);
+        DriverAction.waitSec(3);
+    }
+
```
Comment (gem-rushilmatta) [inline src/test/java/com/qa/helpdesk/stepDefinition/stepDefinition.java:94]: Use try catch block and GemTestReporter here

File: src/test/java/com/qa/helpdesk/stepDefinition/stepDefinition.java [hunk 6]
```diff
@@ -0,0 +1,468 @@
...
+    @And("^Validate user details \"([^\"]*)\"$")
+    public void validateUserDetails(String user) {
+        WebElement val = DriverManager.getWebDriver().findElement(By.xpath(locator.user.replace("replace", user)));
+        String currentUser = val.getText();
+        if (currentUser.equalsIgnoreCase(user)){
+            GemTestReporter.addTestStep("Validate user details", "Successfully validated user: " + user, STATUS.PASS, takeSnapShot());
+        } else {
+            GemTestReporter.addTestStep("Validate user details", "Failed to validate user details", STATUS.FAIL, takeSnapShot());
+        }
+
```
Comment (gem-rushilmatta) [inline src/test/java/com/qa/helpdesk/stepDefinition/stepDefinition.java:134]: Use try catch block here

File: src/test/java/com/qa/helpdesk/stepDefinition/Hook.java [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+        ChromeOptions options = new ChromeOptions();
+
+        options.addArguments("--remote-allow-origins=*");
+        options.addArguments("--incognito");
+        DriverManager.initializeChrome(options);
+        DriverAction.launchUrl(GemJarUtils.getGemJarConfigData("launchUrl"));
+        DriverAction.setImplicitTimeOut(Long.parseLong(GemJarGlobalVar.implicitTime));
+        DriverAction.setPageLoadTimeOut(Long.parseLong(GemJarGlobalVar.pageTimeout));
+        DriverAction.setScriptTimeOut(Long.parseLong(GemJarGlobalVar.scriptTimeout));
+        DriverAction.maximizeBrowser();
```
Comment (Gem-NeerajSingh) [inline src/test/java/com/qa/helpdesk/stepDefinition/Hook.java:23]: There is no need to change the start url and browser if they are already defined in the config file.

File: src/test/java/com/qa/helpdesk/stepDefinition/stepDefinition.java [hunk 8]
```diff
@@ -0,0 +1,468 @@
...
+                    && DriverAction.isExist(locator.SignInPageHelpdesk)
+                    && DriverAction.isExist(locator.SignInIcon)) {
+                GemTestReporter.addTestStep("Verify the Url is launched and Sign in Page is loaded", "Url is launched and page is loaded successfully", STATUS.PASS, takeSnapShot());
+            }
+            else {
+                GemTestReporter.addTestStep("Verify the Url is launched and Sign in Page is loaded", "Failed to launch Url and load page", STATUS.FAIL, takeSnapShot());
+            }
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
```
Comment (Gem-NeerajSingh) [inline src/test/java/com/qa/helpdesk/stepDefinition/stepDefinition.java:40]: Use GemTestReporter  for logging failure as well

File: src/test/java/com/qa/helpdesk/stepDefinition/stepDefinition.java [hunk 9]
```diff
@@ -0,0 +1,468 @@
...
+
+    }
+
+    @When("Sign in to Helpdesk Portal")
+    public void signInToHelpdeskPortal() {
+        DriverAction.click(locator.loginSSO);
+        DriverAction.waitSec(3);
```
Comment (Gem-NeerajSingh) [inline src/test/java/com/qa/helpdesk/stepDefinition/stepDefinition.java:47]: Surrounded by try catch or check visibility of element

File: src/test/java/com/qa/helpdesk/stepDefinition/stepDefinition.java [hunk 10]
```diff
@@ -0,0 +1,468 @@
...
+            DriverAction.waitSec(10);
+            if (isExist(locator.gemLogo) && isExist(locator.myTickets) && (isExist(locator.userGuideBtn))) {
+                GemTestReporter.addTestStep("Verify if the login was successful", "User has logged in to Helpdesk Portal successfully", STATUS.PASS, takeSnapShot());
+            } else {
+                GemTestReporter.addTestStep("Verify if the login was successful", "Failed to login to Helpdesk Portal", STATUS.FAIL, takeSnapShot());
+            }
+        }
+        catch(Exception e) {
+            throw new RuntimeException(e);
+        }
```
Comment (Gem-NeerajSingh) [inline src/test/java/com/qa/helpdesk/stepDefinition/stepDefinition.java:107]: Add GemTestReporter

File: src/test/java/com/qa/helpdesk/stepDefinition/stepDefinition.java [hunk 11]
```diff
@@ -0,0 +1,468 @@
...
+        try {
+            DriverAction.waitUntilElementAppear(locator.myTicketsTable, 15);
+        if (isExist(locator.gemLogo) && isExist(locator.myTickets) && (isExist(locator.userGuideBtn))) {
+            GemTestReporter.addTestStep("Validate Homepage elements are loaded", "Homepage elements are loaded successfully", STATUS.PASS, takeSnapShot());
+        } else {
+            GemTestReporter.addTestStep("Validate Homepage elements are loaded", "Failed to load Homepage elements", STATUS.FAIL, takeSnapShot());
+        }
+        }
+        catch(Exception e) {
+            throw new RuntimeException(e);
```
Comment (Gem-NeerajSingh) [inline src/test/java/com/qa/helpdesk/stepDefinition/stepDefinition.java:121]: Add GemTestReporter

File: src/test/java/com/qa/helpdesk/stepDefinition/stepDefinition.java [hunk 12]
```diff
@@ -0,0 +1,468 @@
...
+            }
+            String rowVal = DriverManager.getWebDriver().findElement(locator.firstRow).getText();
+            if(rowVal != null){
+                GemTestReporter.addTestStep("Validate My Tickets table contains at least 1 row", "Table contains at least 1 row", STATUS.PASS, takeSnapShot());
+            } else {
+            GemTestReporter.addTestStep("Validate My Tickets table contain any row", "Table doesn't contains at least 1 row", STATUS.FAIL, takeSnapShot());
+        }
+        } catch(Exception e) {
+                throw new RuntimeException(e);
+            }
```
Comment (Gem-NeerajSingh) [inline src/test/java/com/qa/helpdesk/stepDefinition/stepDefinition.java:156]: Add GemTestReporter

File: src/test/java/com/qa/helpdesk/stepDefinition/stepDefinition.java [hunk 13]
```diff
@@ -0,0 +1,468 @@
...
+            DriverAction.click(locator.createdOn);
+            String sortedRowVal = DriverManager.getWebDriver().findElement(locator.firstRow).getText();
+            if(rowVal != sortedRowVal) {
+                GemTestReporter.addTestStep("Sorting the rows", "Successfully sorted the rows ", STATUS.PASS, takeSnapShot());
+            } else {
+                GemTestReporter.addTestStep("Sorting the rows", "Failed to sort the rows", STATUS.FAIL, takeSnapShot());
+            }
+
+        } catch(Exception e) {
+            throw new RuntimeException(e);
```
Comment (Gem-NeerajSingh) [inline src/test/java/com/qa/helpdesk/stepDefinition/stepDefinition.java:172]: Add GemTestReporter

File: src/test/java/com/qa/helpdesk/stepDefinition/stepDefinition.java [hunk 14]
```diff
@@ -0,0 +1,468 @@
...
+        }
+    }
+
+    @When("Select Filters")
+    public void selectFilters() {
+
+    }
+
```
Comment (Gem-NeerajSingh) [inline src/test/java/com/qa/helpdesk/stepDefinition/stepDefinition.java:180]: Remove dead code

File: src/test/java/com/qa/helpdesk/stepDefinition/stepDefinition.java [hunk 15]
```diff
@@ -0,0 +1,468 @@
...
+    //Day 2 Create Ticket
+    @Given("^Click on Create New Ticket button$")
+    public void clickOnCreateNewTicketButton() {
+        try {
+            DriverAction.click(locator.createTicketBtn, "Click on Create New Ticket button", "Successfully clicked on Create New Ticket button");
+            DriverAction.waitSec(2);
+
+        } catch (Exception e) {
+            throw new RuntimeException(e);
```
Comment (Gem-NeerajSingh) [inline src/test/java/com/qa/helpdesk/stepDefinition/stepDefinition.java:189]: Add GemTestReporter

File: src/test/java/com/qa/helpdesk/stepDefinition/stepDefinition.java [hunk 16]
```diff
@@ -0,0 +1,468 @@
...
+        try {
+            List <WebElement> createTicketFields = DriverManager.getWebDriver().findElements(locator.ticketCreationDetails);
+            if(createTicketFields.size()==5 && isExist(locator.previewBtn) && (isExist(locator.submitBtn))){
+                GemTestReporter.addTestStep("Verify if Create New Ticket form appears", "Create New Ticket form is present", STATUS.PASS, takeSnapShot());
+            } else {
+                GemTestReporter.addTestStep("Verify if Create New Ticket form appears", "Failed to open Create New Ticket form", STATUS.FAIL, takeSnapShot());
+            }
+
+        } catch (Exception e) {
+            throw new RuntimeException(e);
```
Comment (Gem-NeerajSingh) [inline src/test/java/com/qa/helpdesk/stepDefinition/stepDefinition.java:204]: Add GemTestReporter

File: src/test/java/com/qa/helpdesk/stepDefinition/stepDefinition.java [hunk 17]
```diff
@@ -0,0 +1,468 @@
...
+
+            String cell4 = createTicketFields.get(1).getText();
+            if (cell4.contains("Description")) {
+                DriverAction.click(locator.ticketDescription);
+                DriverAction.typeText(locator.ticketDescription, description);
+                GemTestReporter.addTestStep("Enter Description for the new ticket", "Successfully entered description as: " + description, STATUS.PASS, takeSnapShot());
+            } else {
+                GemTestReporter.addTestStep("Enter Description for the new ticket", "Failed to enter description", STATUS.FAIL, takeSnapShot());
+            }
+
```
Comment (gem-rushilmatta) [inline src/test/java/com/qa/helpdesk/stepDefinition/stepDefinition.java:329]: Switch case can be used to reduce the redundancy of if else statement and common function for repetitive code

File: src/test/java/com/qa/helpdesk/stepDefinition/stepDefinition.java [hunk 18]
```diff
@@ -0,0 +1,468 @@
...
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+
+    @And("^Fill ticket \"([^\"]*)\" \"([^\"]*)\" \"([^\"]*)\" and \"([^\"]*)\" for Incident Type$")
+    public void fillTicketAndForIncidentType(String subject, String incident, String department, String description) throws Throwable {
+
```
Comment (Gem-NeerajSingh) [inline src/test/java/com/qa/helpdesk/stepDefinition/stepDefinition.java:211]: Throwable is superclass of all errors and exceptions , avoid using if using try catch already

File: src/test/java/com/qa/helpdesk/stepDefinition/stepDefinition.java [hunk 19]
```diff
@@ -0,0 +1,468 @@
...
+            }
+
+
+        } catch (Exception e) {
+        throw new RuntimeException(e);
+    }
+    }
+
+    @And("^Fill ticket \"([^\"]*)\" \"([^\"]*)\" \"([^\"]*)\" \"([^\"]*)\" \"([^\"]*)\" and \"([^\"]*)\" for Request Type$")
+    public void fillTicketForRequestType(String subject, String incident, String department, String description, String category, String subCategory) {
```
Comment (Gem-NeerajSingh) [inline src/test/java/com/qa/helpdesk/stepDefinition/stepDefinition.java:284]: \"([^\"]*)\" can be replace with {String} for better understanding and visibility

File: src/test/java/com/qa/helpdesk/stepDefinition/stepDefinition.java [hunk 20]
```diff
@@ -0,0 +1,468 @@
...
+                DriverAction.click(locator.ticketDescription);
+                DriverAction.typeText(locator.ticketDescription, description);
+                GemTestReporter.addTestStep("Enter Description for the new ticket", "Successfully entered description as: " + description, STATUS.PASS, takeSnapShot());
+            } else {
+                GemTestReporter.addTestStep("Enter Description for the new ticket", "Failed to enter description", STATUS.FAIL, takeSnapShot());
+            }
+
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
```
Comment (Gem-NeerajSingh) [inline src/test/java/com/qa/helpdesk/stepDefinition/stepDefinition.java:332]: Add GemTestReporter

File: src/test/java/com/qa/helpdesk/stepDefinition/stepDefinition.java [hunk 21]
```diff
@@ -0,0 +1,468 @@
...
+                DriverAction.waitSec(2);
+                DriverAction.click(locator.continueBtn);
+                DriverAction.waitSec(5);
+                GemTestReporter.addTestStep("Verify if ticket is created successfully", "" + creationSuccess + "with Id: " + ticketID, STATUS.PASS, takeSnapShot());
+            } else {
+                GemTestReporter.addTestStep("Verify if ticket is created successfully", "Failed to create new ticket", STATUS.PASS, takeSnapShot());
+            }
+
+        } catch (Exception e) {
+            throw new RuntimeException(e);
```
Comment (Gem-NeerajSingh) [inline src/test/java/com/qa/helpdesk/stepDefinition/stepDefinition.java:351]: Add GemTestReporter

File: src/test/resources/features/feature.feature [hunk 22]
```diff
@@ -0,0 +1,91 @@
...
+#  Scenario Outline: Create New Ticket Functionality for Request
+#    Given Click on Create New Ticket button
+#    When Create New Ticket dialog box should appear
+#    And Fill ticket "<subject>" "<type>" "<department>" "<description>" "<category>" and "<subCategory>" for Request Type
+#    Then Preview ticket details and click on Submit button
+#
+#    Examples:
+#      | subject | type | department | description | category | subCategory |
+#      | New Hardware requirement | Request | IT | This is test| Hardware | Monitor |
+
```
Comment (Gem-NeerajSingh) [inline src/test/resources/features/feature.feature:49]: Remove dead code

**Review by gem-rushilmatta — CHANGES_REQUESTED** [review]
Make changes as per the given comments

**Pull request closed without merging**